### ISSUE TITLE
Wrap logs from java integrations in AirbyteLogMessage

### ISSUE THREAD

**Issue by cgardens** [airbytehq/airbyte#659]
## Tell us about the problem you're trying to solve
The `java-base` module is a lib used to write Java-based connectors. `AirbyteLogMessage` did not exist when java base was written, which mean the logs output here are not wrapped in a structured json message as described in the protocol ([link](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml#L32)).

This means there are connectors logs that are are being output to standard out that the platform is permissively accepting. Modifying base-java will help clean this up and allow people writing connectors in java to get that functionality "for free.

This only applies to all connector related code pieces in the module.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368111390764) by [Unito](https://www.unito.io)

**Comment by michel-tricot** [comment]
I wonder if that can be made automatically with log4j/slf4j

**Comment by cgardens** [comment]
🤞 yeah. I hope so would be nice to do it using that then having to write a custom logger.

**Comment by davinchia** [comment]
I think this is already done?

**Comment by cgardens** [comment]
Nope. The platform is permissive, so if it receives non-json it assumes it is a log. But ideally our java-base would actually follow the protocol and emit structure logs. Maybe a good first issue for someone who is coming in with existing java XP?

**Comment by davinchia** [comment]
I'm confused. Is it this [base-java](https://github.com/airbytehq/airbyte/tree/master/airbyte-integrations/bases/base-java) module?

My understanding is structured logs are for the connectors. Everything here seems to be within the platform/workers. What am I missing?

**Comment by cgardens** [comment]
Yes. I'm referring to the base-java module that you linked. And yes, we are just talking about connectors. That is a library to build off of for connectors written in java. Instead of just logging log lines to stdout, it should log them as a structure json long message as described in the protocol ([link](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml#L30-L32
)). Likely the team best situated to work on this would be dbs & dws since they are the ones who write connectors in java. 

Does that clear it up?

**Comment by davinchia** [comment]
Yeap that does, thanks! My confusion was I thought I saw more platform code than connector code here. On closer observation, it's definitely the other way round.

**Comment by cgardens** [comment]
:+1: should be exclusively connector code in that module.

**Comment by davinchia** [comment]
Updated the description, thanks Charles!

**Comment by yurii-bidiuk** [comment]
Hi @cgardens
Something from the description is not clear for me to understand the acceptance criteria for this task.
As far as I know, now we log messages to stdout by slf4j Logger (e.g. `log.info("some message here");`)

> instead of just logging log lines to stdout, they should log them as a structure JSON long message as described in the protocol

Do you mean continue logging to stdout, but using the structure of `AirbyteLogMessage` and log message will be in JSON format (`{"level":"info","message":"some message here"}`)? Or you mean using another way of logging?
Could you please clarify?
Thanks

**Comment by cgardens** [comment]
@yurii-bidiuk  yeah. that's exactly what i mean.

example:
`log.info("some message here");`
=>
`{ "type": "LOG", "log": { "level":"info","message":"some message here" }`

this should be how all java connectors log, but not the rest of the platform. does that help?

**Comment by yurii-bidiuk** [comment]
@cgardens  Yes, it helped me a lot. Thank you!

I'm double-checking to avoid misunderstanding:
Now some log message from the connector has the next view:
`2022-07-21 10:03:35 INFO i.a.i.s.j.AbstractJdbcSource(lambda$discoverInternal$5):145 - Table models column model (type varchar[200]) -> JsonSchemaType({type=string})`

After implementation of this thick a log would be like:
`2022-07-21 10:03:35 INFO i.a.i.s.j.AbstractJdbcSource(lambda$discoverInternal$5):145 - { "type": "LOG", "log": { "level":"info","message":"Table models column model (type varchar[200]) -> JsonSchemaType({type=string})" }`

Is my understanding correct?

**Comment by cgardens** [comment]
That doesn't sound right to me. I would imagine it would be something like this. @evantahler or @jdpgrailsdev want to weight in here? I'm probably not the right person to specify the exact format.

```
{ "type": "LOG", "log": { "level":"info","message":"2022-07-21 10:03:35 INFO i.a.i.s.j.AbstractJdbcSource(lambda$discoverInternal$5):145 - Table models column model (type varchar[200]) -> JsonSchemaType({type=string}) }}"
```

**Comment by jdpgrailsdev** [comment]
> That doesn't sound right to me. I would imagine it would be something like this.
> 
> `{ "type": "LOG", "log": { "level":"info","message":"2022-07-21 10:03:35 INFO i.a.i.s.j.AbstractJdbcSource(lambda$discoverInternal$5):145 - Table models column model (type varchar[200]) -> JsonSchemaType({type=string}) }}"`
> 
> @evantahler or @jdpgrailsdev want to weight in here? I'm probably not the right person to specify the exact format.

@cgardens I don't have strong feelings here either.  On one hand, not tying the "message" to the SLF4J format string might be advantageous to avoid confusion if logged also by the platform.  However, as the intent appears to be the storage of logs emitted by the connector for display by the platform, it is probably best for it to be opaque to the protocol.  That being said, I do wonder if having the level is a bit redundant, unless we intend to have our own Airbyte log levels different from the SLF4J log levels, for whatever reason.  I suppose that we may need to do some thinking on this to figure out what data we need to capture as part of a connector log message.

**Comment by evantahler** [comment]
This is a cool project, which might benefit from a mini-PRD to think though the edge cases and goals.  Here's what I think the goals are:

1. Every message emitted by a connector should /eventually/ be JSON that conforms to the spec - this issue is a step on that journey.  This is valuable because one day we can build parsing/filtering features in the UI (e.g. show me all messages with a level >= "info") as @jdpgrailsdev hints at above, or maybe drop all "debug" messages, etc.
3. For some well-structured message formats (like SLF4J) we may be able to infer some of the metadata we need (like "level"), or we can use a CDK or shared library to emit messages in the proper format. 

Wit this in mind, I think my suggestions for guiding principals for this would would be:
1. All messages should be valid JSON at the line-level. This agrees with @cgardens'note above that 

```js
{ "type": "LOG", "log": { "level":"info","message":"2022-07-21 10:03:35 INFO i.a.i.s.j.AbstractJdbcSource(lambda$discoverInternal$5):145 - Table models column model (type varchar[200]) -> JsonSchemaType({type=string}) }}
``` 

is preferable to 

```js
2022-07-21 10:03:35 INFO i.a.i.s.j.AbstractJdbcSource(lambda$discoverInternal$5):145 - { "type": "LOG", "log": { "level":"info","message":"Table models column model (type varchar[200]) -> JsonSchemaType({type=string})" }
```

Take care to remove or serialize line breaks and other special chars so that the produced JSON message remains valid. 

2. We probably should store the original messages with as much richness as we can.  Ideally, we don't need to duplicate information like "level" or "timestamp" if we control the whole logging  pipeline:

Preferred:
```
log.info("some message here");
=>
{ "type": "LOG", "log": { "level":"info","message":"some message here" }
``` 

...but if we don't form the message content ourselves (e.g. it comes from a library), duplicating data is OK. For example, "level" being repeated again within the message content itself is better than accidentally loosing it.

**Comment by yurii-bidiuk** [comment]
@cgardens @jdpgrailsdev @evantahler 
Thank you guys for a detailed explanation. 

**Acceptance criteria** for this ticket would be:
- In the `java-base` module implement a custom logger to log all connectors logs as valid JSON at the line-level;
- JSON structure for such logs should have the next view: 
`{ "type": "LOG", "log": { "level":"info","message":"some message here" }`;
- Ideally, we don't need to duplicate information like "level" or "timestamp", but if we don't form the message content ourselves (e.g. it comes from a library), duplicating data is OK;

 _(open question)_  what should we do with existing connector logs? Should all of them be written by our custom logger and have a JSON structure? For example, in `PostgresSource.discoverRawTables()` we log messages using SLF4J logger:
![Screenshot from 2022-08-04 13-11-12](https://user-images.githubusercontent.com/35812734/182823757-1e03bb4b-20e7-42f2-9b65-faf95f2f0c98.png)
 From my understanding, the SLF4J logger should be replaced with our custom logger. Am I right?

**Comment by grishick** [comment]
I don't think we have to get rid of SLA4J. A more flexible approach could be to implement our own Logger and LoggerFactory that will be the default and allow OSS users to configure their own LoggerFactory and adapter in case they want to log into ELK stack or somewhere else. Some more info here: https://www.slf4j.org/faq.html#slf4j_compatible

**Comment by yurii-bidiuk** [comment]
We need to implement our custom logger described [here](https://github.com/airbytehq/airbyte/issues/659#issuecomment-1205054681) based on SLF4J API. It will be a logger adapter like [Reload4jLoggerAdapter](https://www.slf4j.org/apidocs//src-html/org/slf4j/reload4j/Reload4jLoggerAdapter.html) or [JDK14LoggerAdapter](https://github.com/qos-ch/slf4j/blob/master/slf4j-jdk14/src/main/java/org/slf4j/jul/JDK14LoggerAdapter.java). For such adapter we need to create custom LoggerFactory like [Reload4jLoggerFactory](https://www.slf4j.org/apidocs/src-html/org/slf4j/reload4j/Reload4jLoggerFactory.html) or [JDK14LoggerAdapter](https://github.com/qos-ch/slf4j/blob/master/slf4j-jdk14/src/main/java/org/slf4j/jul/JDK14LoggerAdapter.java).

**Comment by grishick** [comment]
> @davinchia @cgardens @grishick @evantahler Hello team! By adding Log4j2 and JsonTemplateLayout config to java-base modulle I wrapped all connector's logs to [JSON structure ](https://github.com/airbytehq/airbyte/pull/15668#issuecomment-1215611228) as described above. You can find it in [this PR](https://github.com/airbytehq/airbyte/pull/15668). If we run [integrations tests](https://github.com/airbytehq/airbyte/runs/7856479876?check_suite_focus=true) for some connector, all connector's logs will be JSON structured. But if we run sync from UI, all JSON logs are rewrapped to simple string message by [this method](https://github.com/airbytehq/airbyte/blob/7b837180c1891fbeb47ff79593f57da303f63ee1/airbyte-workers/src/main/java/io/airbyte/workers/internal/DefaultAirbyteStreamFactory.java#L86): ![Screenshot from 2022-08-16 17-39-02](https://user-images.githubusercontent.com/35812734/184908294-a27f34d2-9cb8-4bf8-996f-391518a57f8e.png)
> 
> Could you tell me please is that expected behavior?

AFAIK, this is the expected behavior

**Comment by cgardens** [comment]
@evantahler can you confirm that this will work well for the platform?

**Comment by evantahler** [comment]
Here's a sample from the integration test linked above:

```
    ***"type":"LOG","timestamp":"2022-08-16T11:18:18.816Z","log":***"level":"INFO","message":"using CDC: true"***
    ***"type":"LOG","timestamp":"2022-08-16T11:18:18.816Z","log":***"level":"INFO","message":"Set initial fetch size: 10 rows"***
    ***"type":"LOG","timestamp":"2022-08-16T11:18:18.820Z","log":***"level":"INFO","message":"For CDC, only tables in publication publication will be included in the sync: [models_schema.models, models_schema_random.models_random]"***
    ***"type":"LOG","timestamp":"2022-08-16T11:18:18.820Z","log":***"level":"INFO","message":"using CDC: true"***
    ***"type":"LOG","timestamp":"2022-08-16T11:18:18.820Z","log":***"level":"INFO","message":"First record waiting time: 5 seconds"***
    ***"type":"LOG","timestamp":"2022-08-16T11:18:18.820Z","log":***"level":"INFO","message":"First record waiting time: 5 seconds"***
    ***"type":"LOG","timestamp":"2022-08-16T11:18:18.820Z","log":***"level":"INFO","message":"Attempting to find the named replication slot using the query: HikariProxyPreparedStatement@57151587 wrapping SELECT * FROM pg_replication_slots WHERE slot_name = 'debezium_slot_db_qjqqpwtqge' AND plugin = 'pgoutput' AND database = 'db_qjqqpwtqge'"***
    ***"type":"LOG","timestamp":"2022-08-16T11:18:18.820Z","log":***"level":"INFO","message":"Set initial fetch size: 10 rows"***
    ***"type":"LOG","timestamp":"2022-08-16T11:18:18.828Z","log":***"level":"INFO","message":"StandaloneConfig values: \n\taccess.control.allow.methods = \n\taccess.control.allow.origin = \n\tadmin.listeners = null\n\tbootstrap.servers = [localhost:9092]\n\tclient.dns.lookup = use_all_dns_ips\n\tconfig.providers = []\n\tconnector.client.config.override.policy = All\n\theader.converter = class org.apache.kafka.connect.storage.SimpleHeaderConverter\n\tkey.converter = class org.apache.kafka.connect.json.JsonConverter\n\tlisteners = [http://:8083]\n\tmetric.reporters = []\n\tmetrics.num.samples = 2\n\tmetrics.recording.level = INFO\n\tmetrics.sample.window.ms = 30000\n\toffset.flush.interval.ms = 1000\n\toffset.flush.timeout.ms = 5000\n\toffset.storage.file.filename = /tmp/cdc-state-offset18269736122155646632/offset.dat\n\tplugin.path = null\n\tresponse.http.headers.config = \n\trest.advertised.host.name = null\n\trest.advertised.listener = null\n\trest.advertised.port = null\n\trest.extension.classes = []\n\tssl.cipher.suites = null\n\tssl.client.auth = none\n\tssl.enabled.protocols = [TLSv1.2, TLSv1.3]\n\tssl.endpoint.identification.algorithm = https\n\tssl.engine.factory.class = null\n\tssl.key.password = null\n\tssl.keymanager.algorithm = SunX509\n\tssl.keystore.certificate.chain = null\n\tssl.keystore.key = null\n\tssl.keystore.location = null\n\tssl.keystore.password = null\n\tssl.keystore.type = JKS\n\tssl.protocol = TLSv1.3\n\tssl.provider = null\n\tssl.secure.random.implementation = null\n\tssl.trustmanager.algorithm = PKIX\n\tssl.truststore.certificates = null\n\tssl.truststore.location = null\n\tssl.truststore.password = null\n\tssl.truststore.type = JKS\n\ttask.shutdown.graceful.timeout.ms = 5000\n\ttopic.creation.enable = true\n\ttopic.tracking.allow.reset = true\n\ttopic.tracking.enable = true\n\tvalue.converter = class org.apache.kafka.connect.json.JsonConverter\n"***
```

I'm most concerned with how things look in the UI at the moment, but it looks like we have a good way to format messages for the user already, so I'm 👍 

Adding @sherifnada just in case the SAT tests expect to see log messages in a certain way (I don't think so).

In some future project, it would be great to shrink/hide the `i.a.w.i.DefaultAirbyteStreamFactory(...)` prefix when displaying logs in the UI

One final thing to check @yurii-bidiuk - can you also attach a downloaded file of a sync's log?   Ideally, that file's contents are formatted the same as the UI

**Comment by yurii-bidiuk** [comment]
@evantahler 
Here is a file of a sync's log. I made sync with error to show you how connector's exception is logged. As you can see from logs or from [our discussion](https://github.com/airbytehq/airbyte/pull/15668#issuecomment-1220099320) with @edgao, the original stack trace is logged only as part of `sync summary` log
[logs-1388.txt](https://github.com/airbytehq/airbyte/files/9401607/logs-1388.txt)

**Comment by edgao** [comment]
this has apparently regressed, i.e. connectors are once again logging raw text to stdout https://airbytehq-team.slack.com/archives/C02UF50V9HA/p1700274101207249

**Comment by davinchia** [comment]
Note: if you look at the log4j config, the previous json appenders are no longer there.
